Several OsRng improvements

pitdicker opened the pull request:
- Documentation links to the relevant man pages.
- A bit more explanation around wasm, as it turns out to be a bit confusing.
- Removed the blanket Unix implementation of reading from `/dev/urandom`. Besides Linux (which it is pretty specialized for), it was only 'correct' for NetBSD. I think it is better to not compile, and add support for a platform, than to have a default implementation. Then we can make sure it is tested, and matches the recommendations for that platform.
- NetBSD: now uses a simple zero-sized struct for `OsRng`.
- Dragonfly BSD: reads from `/dev/random`. `/dev/urandom` is faster, but using IBAA, the predecessor of ISAAC, of which the security is unknown.
- Added support for Bitrig. It is obsolete, but costs us nothing, as it is the same as OpenBSD.
- Emscripten: reads from `/dev/urandom`, without testing `/dev/random`, and splits the read up in chunks.
- `stdweb` split the read up in chunks of 65536 bytes.
- Moved shared code around reading from a random device to a `unix` module.
- Removed the WASM without `stdweb` stub (never understood why we should have it).
- Added test that reading 0 bytes works.

Fixes https://github.com/rust-lang-nursery/rand/issues/332.

pitdicker commented:
> Did you test this for all those operating systems?

I wish :smile:. But most of the changes are around `/dev/random` and `/dev/urandom`. I did test those changes by faking the module to be Linux, and made the choice between those, or to use chunks, from the man pages.

I finished digging though the Haiku source code (with help from a virtual machine). There is just about no documentation, but `/dev/random` and `/dev/urandom` seem to be [the same](https://github.com/haiku/haiku/blob/master/src/add-ons/kernel/bus_managers/random/driver.cpp#L207), and it uses the [Yarrow RNG](https://github.com/haiku/haiku/blob/master/src/add-ons/kernel/bus_managers/random/yarrow_rng.cpp). So added support for that.

And I am trying to test Solaris with a syscall interface. Trying to cross-compile from a virtual machine with Ubuntu, to test it on a virtual machine with Solaris. :cry:

dhardy commented on src/rngs/os.rs:
```diff
@@ -305,85 +405,81 @@ mod imp {
-        Ok(())
+    impl OsRng {
+        pub fn new() -> Result<OsRng, Error> {
+            open_random_device("/dev/urandom", false)?;
```
Comment above says read from `/dev/random`. Which is correct?

pitdicker commented:
> Initially this was added to allow compiling to WASM, possibly with the goal of making remaining functionality available (not quite sure)

I think that explaines it. We didn't have a `no_std` mode at the time, so that only the traits and PRNGs could be used, and also no `rand_core` crate.

> Regardless, I'm a little concerned about this change, especially for the 0.5 release, though it may be the best option in the end.

I am not completely used to the idea of having some sort of stability :smile:. But I don't think this PR really makes anything worse on the systems that Rust supports.

> (Also that snippet in JitterRng which just fails — ideally we would not include features which just fail at run-time.)

I think we need to test someday that is using a good timer on all platforms Rust supports. And then we may even re-implement that little part that `std` takes care of now ourselves to. But I am not sure what to do if there is no good timer available, as with WASM.

pitdicker commented:
I am not sure adding the commit to add support for Solaris is the right thing to add.

I think it is correct or pretty close, as it is mostly the same as the one for Linux. It compiles, but I just can't get it to link against my copy of the libraries from Solaris...

Is having something better than nothing?

dhardy commented on src/rngs/os.rs:
```diff
@@ -483,6 +488,127 @@ mod imp {
+                } else if kind == io::ErrorKind::WouldBlock {
+                    // Potentially this would waste bytes, but since we use
+                    // /dev/urandom blocking only happens if not initialised.
+                    // Also, wasting the bytes in dest doesn't matter very much.
```
Did you understand the [man page](https://docs.oracle.com/cd/E86824_01/html/E54765/getrandom-2.html)? Apparently the function returns -1 when it would block, and may set 0 or the number of bytes requested, and one should check the return value — but this is -1!

pitdicker replied:
That is an old comment, I noticed it after making the commit. I was going over the documentation a bit more right now.

pitdicker replied:
Hmm, the manual is not very clear. Both -1 and 0 can indicate an error. We should ask for bytes in chunks of 1024, while it are chunks of 1040 for `/dev/random`. And apparently it is guaranteed the chunk it either not or completely filled.

dhardy replied:
1040 is an unusual chunk size. Yeah, I love confusing manuals (not).

pitdicker commented:
Sorry, I keep playing with this file.

I have tried to refactor the `OsRng` implementations, using this trait:
```rust
trait OsRngImpl where Self: Sized {
    fn new() -> Result<Self, Error>;
    fn fill_chunk(&mut self, dest: &mut [u8]) -> Result<(), Error>;
    fn max_chunk_size(&self) -> Option<usize> { None }
    fn method_str(&self) -> &'static str;
}
```

In my opinion it helped making things easier to understand. Logging and dividing into chunks is now done in one place (and thus uniformly), in `try_fill_bytes`.

I have tried cross-compiling almost all targets. It turned out Fuchsia and CloudABI didn't work before...

Redox doesn't work yet. The `unix` module depends on `libc`, for test opening `/dev/random`. Maybe it is best to move that piece elsewhere, but not sure how yet. It is quite platform-specific.

dhardy commented on src/rngs/os.rs:
```diff
@@ -188,10 +188,31 @@ impl RngCore for OsRng {
+trait OsRngImpl where Self: Sized {
+    fn new() -> Result<Self, Error>;
+    fn fill_chunk(&mut self, dest: &mut [u8]) -> Result<(), Error>;
+    fn max_chunk_size(&self) -> Option<usize> { None }
```
May as well just use `core::usize::MAX` and drop the `Option` since this is the default impl.

dhardy commented on src/rngs/os.rs:
```diff
@@ -518,11 +552,25 @@ mod imp {
-                OsRngMethod::RandomDevice => read_random_device(slice, Some(1040)),
-             }
+                OsRngMethod::GetRandom => Some(1024),
+                OsRngMethod::RandomDevice => Some(1040),
```
Seriously? Save the logic and just use 1024.

dhardy commented on src/rngs/os.rs:
```diff
@@ -834,37 +877,31 @@ mod imp {
         }
+
+        fn max_chunk_size(&self) -> Option<usize> {
+            Some(<ULONG>::max_value() as usize)
```
I assume this is `u32::MAX`. Should probably be the default; it's not like anyone will use more anyway.

pitdicker commented on src/rngs/os.rs:
```diff
@@ -539,20 +542,21 @@ mod imp {
+        }
+        for slice in dest.chunks_mut(max) {
+            let result = getrandom(&mut slice);
+            if result == -1 || result == 0 {
```
I think in this case I like to point out both, so it is clearly not a typo.

pitdicker commented:
I am mostly happy with it now. And thank you for also looking carefully :smile.

I am not sure yet what the right thing to do is around blocking. On Linux, NetBSD and Solaris we have the choice between an error and blocking. Other systems always block, or don't document what happens when the OS RNG hasn't yet collected enough entropy.

And Linux, FreeBSD (not NetBSD) and Solaris have the ability to poll on `/dev/random` to wait until it becomes available.

We currently choose to get an error, and have some approximation of blocking in the `fill_bytes` wrapper using `thread::sleep()`.

Would it make sense to store whether we know the OS RNG is not yet initialized, and to use a blocking read in `fill_bytes` instead of trying to schedule ourselves? And to expose some functions like this for users?

pitdicker commented:
Added one more commit. This is not getting easier...

I added a `test_initialized` methods. On most systems this is a no-op. On Linux, NetBSD and Solaris is can be used to return an error if the OS RNG it not yet initialized, or to block instead. Once we know the OS RNG is initialized, this is stored in a static and not checked again.

`fill_bytes` will use the blocking variant, `try_fill_bytes` the error variant. A small advantage over doing the check in `OsRng::new()` is that the random bytes generated are not wasted, and that we don't have to do our own primitive scheduling.

pitdicker commented:
Update to use the correct open flags on Solaris.

dhardy commented:
I might wait a couple of days before reviewing just in case :laughing:
